# Ejercicios Pair Programming 
## Lección 7 - ANOVA

En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de normalización y estandarización.

In [22]:
#Librerias

import pandas as pd
from scipy import stats
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Warning
import warnings
warnings.filterwarnings('ignore')

In [23]:
df = pd.read_csv("supermercado_normalizado.csv", index_col=0)
df_standard = pd.read_csv("supermercado_estandarizado_standard.csv", index_col=0)
df_robust = pd.read_csv("supermercado_estandarizado_robust.csv", index_col=0)

In [24]:
df.head()

id  year_birth   education marital_status   income dt_customer  \
0   1826        1970  Graduation       Divorced  84835.0   6/16/2014   
1      1        1961  Graduation         Single  57091.0   6/15/2014   
2  10476        1958  Graduation        Married  67267.0   5/13/2014   
3   1386        1967  Graduation       Together  32474.0   11/5/2014   
4   5371        1989  Graduation         Single  21474.0    8/4/2014   

   numdealspurchases  numstorepurchases  numwebvisitsmonth  \
0                  1                  6                  1   
1                  1                  7                  5   
2                  1                  5                  2   
3                  1                  2                  7   
4                  2                  2                  7   

   numwebpurchases_sklearn  
0                 0.363636  
1                 0.636364  
2                 0.272727  
3                 0.090909  
4                 0.272727

In [25]:
df_standard.head()

id  year_birth   education marital_status    income dt_customer  \
0 -1.157073    0.092613  Graduation       Divorced  1.608976   6/16/2014   
1 -1.719067   -0.676627  Graduation         Single  0.264394   6/15/2014   
2  1.506626   -0.933040  Graduation        Married  0.757562   5/13/2014   
3 -1.292567   -0.163800  Graduation       Together -0.928642   11/5/2014   
4 -0.065418    1.716564  Graduation         Single -1.461745    8/4/2014   

   numdealspurchases  numstorepurchases  numwebvisitsmonth  \
0          -0.701068           0.052191          -1.803407   
1          -0.701068           0.361131          -0.142553   
2          -0.701068          -0.256748          -1.388194   
3          -0.701068          -1.183568           0.687874   
4          -0.170569          -1.183568           0.687874   

   numwebpurchases_sklearn  
0                 0.363636  
1                 0.636364  
2                 0.272727  
3                 0.090909  
4                 0.272727

In [26]:
df_robust.head()

id  year_birth   education marital_status    income dt_customer  \
0 -0.647571    0.000000  Graduation       Divorced  1.014008   6/16/2014   
1 -0.973274   -0.500000  Graduation         Single  0.175079   6/15/2014   
2  0.896176   -0.666667  Graduation        Married  0.482783   5/13/2014   
3 -0.726096   -0.166667  Graduation       Together -0.569295   11/5/2014   
4 -0.014902    1.055556  Graduation         Single -0.901915    8/4/2014   

   numdealspurchases  numstorepurchases  numwebvisitsmonth  \
0               -0.5                0.2              -1.25   
1               -0.5                0.4              -0.25   
2               -0.5                0.0              -1.00   
3               -0.5               -0.6               0.25   
4                0.0               -0.6               0.25   

   numwebpurchases_sklearn  
0                 0.363636  
1                 0.636364  
2                 0.272727  
3                 0.090909  
4                 0.272727

Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.

📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

De los tres dataframes que tenemos (normalizado, estandarizado con el método Standard y estandarizado con el método Robust) hemos decidido hacer el estudio con el dataframe estandarizado con el método Standard ya que hasta que no hagamos la regesión lineal no podemos llegar a conclusiones fiables. 

En el caso de que no lleguemos a conclusiones satisfactorias con este dataframe, podríamos volver a realizar el estudio de Anova con los otros dataframes.

### ANOVA del dataframe estandarizado con el método Standard

In [27]:
# Iniciamos nuestro anova. Observamos que nuestras variables predictoras son de efecto no aditivos, es decir,
# entendemos que no se influyen mutuamente (son independientes) a la hora de tomar la decisión de hacer la compra online.
lm = ols('numwebpurchases_sklearn ~ id + year_birth + education + marital_status + income + dt_customer + numdealspurchases + numstorepurchases + numwebvisitsmonth' , data=df_standard).fit()
sm.stats.anova_lm(lm)

df     sum_sq    mean_sq           F         PR(>F)
education             4.0   0.127599   0.031900    1.197063   3.102111e-01
marital_status        7.0   0.452039   0.064577    2.423298   1.808282e-02
dt_customer         661.0  37.077203   0.056093    2.104912   6.287032e-32
id                    1.0   0.190617   0.190617    7.153037   7.565626e-03
year_birth            1.0   1.504337   1.504337   56.451242   9.889321e-14
income                1.0  21.392387  21.392387  802.763719  1.307161e-141
numdealspurchases     1.0  10.928258  10.928258  410.090232   9.774733e-81
numstorepurchases     1.0   2.666147   2.666147  100.048955   7.674496e-23
numwebvisitsmonth     1.0   9.216617   9.216617  345.859747   1.394086e-69
Residual           1489.0  39.679501   0.026648         NaN            NaN

Cuando realizamos nuestra anova en el dataframe estandarizado con el método "Standard" vemos que los grados de libertad en las columnas categóricas es en education "4.0" (hay 5 valores únicos), marital_status "7.0" (hay 8 valores únicos), dt_customer "661.0" (hay 662 valores únicos). En el caso de las variables numéricas vemos que el resultado de todas es 1, ya que solo pueden variar dentro de un eje.

Un dato que nos llama la atención es el valor de dt_customer (661.0).

Analizando el PR(<F), es decir el "p-valor" debemos tener en cuenta que nuestra hipótesis nula es que las variables predictoras no influyen en la compra online.
Teniendo en cuenta esto observamos que "education" tiene un valor mayor que 0.05, por lo que no influye en la variable respuesta, es decir, en las compras online. Todo el resto de las variables tiene un valor menor a 0.05, por lo que si influyen en la variable respuesta.

Una vez hecho este estudio lo confirmamos a través de un summary.

In [28]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                               OLS Regression Results                              
===================================================================================
Dep. Variable:     numwebpurchases_sklearn   R-squared:                       0.678
Model:                                 OLS   Adj. R-squared:                  0.532
Method:                      Least Squares   F-statistic:                     4.636
Date:                     Mon, 23 Jan 2023   Prob (F-statistic):          4.88e-134
Time:                             10:39:03   Log-Likelihood:                 1258.4
No. Observations:                     2166   AIC:                            -1163.
Df Residuals:                         1489   BIC:                             2683.
Df Model:                              676                                         
Covariance Type:                 nonrobust                                         
==============================================================================================
                                 coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------------
Intercept                      0.6195      0.158      3.933      0.000       0.311       0.929
education[T.Basic]            -0.0267      0.032     -0.842      0.400      -0.089       0.036
education[T.Graduation]       -0.0045      0.016     -0.279      0.780      -0.036       0.027
education[T.Master]           -0.0003      0.018     -0.016      0.988      -0.036       0.035
education[T.PhD]               0.0017      0.018      0.097      0.923      -0.033       0.036
marital_status[T.Alone]       -0.0940      0.169     -0.557      0.578      -0.425       0.237
marital_status[T.Divorced]    -0.0684      0.134     -0.510      0.610      -0.331       0.195
marital_status[T.Married]     -0.0682      0.134     -0.510      0.610      -0.330       0.194
marital_status[T.Single]      -0.0756      0.134     -0.565      0.572      -0.338       0.187
marital_status[T.Together]    -0.0684      0.134     -0.511      0.610      -0.331       0.194
marital_status[T.Widow]       -0.0737      0.136     -0.543      0.587      -0.340       0.193
marital_status[T.YOLO]        -0.4543      0.201     -2.263      0.024      -0.848      -0.061
dt_customer[T.1/1/2014]       -0.3927      0.142     -2.762      0.006      -0.672      -0.114
dt_customer[T.1/10/2012]      -0.2557      0.110     -2.321      0.020      -0.472      -0.040
dt_customer[T.1/10/2013]      -0.0145      0.125     -0.116      0.907      -0.259       0.230
dt_customer[T.1/11/2012]      -0.1407      0.142     -0.992      0.321      -0.419       0.137
dt_customer[T.1/11/2013]      -0.2123      0.183     -1.160      0.246      -0.571       0.147
dt_customer[T.1/12/2012]      -0.1587      0.116     -1.364      0.173      -0.387       0.069
dt_customer[T.1/12/2013]      -0.1790      0.110     -1.621      0.105      -0.396       0.038
dt_customer[T.1/13/2013]      -0.2220      0.106     -2.101      0.036      -0.429      -0.015
dt_customer[T.1/13/2014]      -0.2468      0.183     -1.349      0.178      -0.606       0.112
dt_customer[T.1/14/2013]      -0.1636      0.183     -0.894      0.371      -0.522       0.195
dt_customer[T.1/14/2014]      -0.2522      0.142     -1.779      0.076      -0.530       0.026
dt_customer[T.1/15/2013]      -0.2149      0.142     -1.513      0.131      -0.494       0.064
dt_customer[T.1/15/2014]      -0.1342      0.116     -1.154      0.249      -0.362       0.094
dt_customer[T.1/16/2013]      -0.1713      0.103     -1.665      0.096      -0.373       0.030
dt_customer[T.1/17/2013]      -0.2182      0.106     -2.064      0.039      -0.426      -0.011
dt_customer[T.1/17/2014]      -0.1680      0.116     -1.451      0.147      -0.395       0.059
dt_customer[T.1/18/2014]      -0.1315      0.125     -1.048      0.295      -0.378 

Analizando el "summary" lo primero que observamos es el R2 (tiene un valor de 0.678). Al ser el valor del R2 elevado sabemos que hemos utilizado unas variables predictoras para nuestro modelo correctas, ya que el cambio de nuestra variable respuesta casi un 70% depende de nuestras variables predictoras.

Observamos a la hora de realizar el "summary" que en las variables predictoras id y year_birth el p-value es completamente diferente (sus valores en uno y otro estudio nos reflejan respuestas contradictorias). Nos quedaremos con el p-value que nos aparece en el summary. En este caso el valor es superior a 0.05, por lo tanto no tiene efecto en la variable respuesta.

En cuanto al parámetro "coef" vemos que tanto "income", "numdealspurchases", "numstorepurchases" y "numwebvisitsmonth" tienen resultados positivos por lo que influye directamente en la compra online, es decir, a mayor ingresos, ofertas, compras en tiendas y número de visitas, mayor es el número de compras online.

En conclusión, vemos que los principales valores que influyen en la compra online según nuestro estudio son: "income", "numdealspurchases", "numstorepurchases", "numwebvisitsmonth", "marital_status" y "dt_customer" no influyendo todos por igual en la variable respuesta (lo que nos puede llevar a eliminar alguna variable predictora en un futuro). 

Hemos decidido eliminar dt_customer del estudio por su alto número de valores únicos.

In [29]:
df_standard.drop(["year_birth", "id", "education","dt_customer"], axis=1, inplace=True)

In [30]:
df_standard.head()

marital_status    income  numdealspurchases  numstorepurchases  \
0       Divorced  1.608976          -0.701068           0.052191   
1         Single  0.264394          -0.701068           0.361131   
2        Married  0.757562          -0.701068          -0.256748   
3       Together -0.928642          -0.701068          -1.183568   
4         Single -1.461745          -0.170569          -1.183568   

   numwebvisitsmonth  numwebpurchases_sklearn  
0          -1.803407                 0.363636  
1          -0.142553                 0.636364  
2          -1.388194                 0.272727  
3           0.687874                 0.090909  
4           0.687874                 0.272727

In [31]:
df_standard.to_csv("supermercado_anova.csv")

Happy coding💪🏽